File: zigzag_conversion/zigzag_conversion.py
#!/usr/bin/env python3
# -*- conding: utf-8 -*-

class Solution:
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        l = len(s);
        if l == 0 or numRows <= 1:
            return s
        row = ["" for i in range(numRows)]
        step = 1
        cnt = 0
        ans = ""
        for i in range(l):
            if cnt == 0:
                step = 1
            if cnt == numRows - 1:
                step = -1
            row[cnt] += s[i]
            cnt += step
        for x in row:
            ans += x
        return ans

if __name__ == '__main__':
    s = Solution()
    print(s.convert("PAYPALISHIRING", 3))
    print(s.convert("PAYPALISHIRING", 4))
    print(s.convert("P", 1))
    print(s.convert("PAYPALISHIRING", 1))
    print(s.convert("", 3))

File: zigzag_conversion/zigzag_conversion.cpp
#include<iostream>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        int len = s.size();
        if(len == 0)return s;
        int cnt = 2*numRows - 2;
        if(numRows == 1)cnt = 1;
        int grp = (len % cnt != 0)?len/cnt+1:len/cnt;
        char ans[len+1];
        int total = 0;
        for(int i=1; i<=numRows; i++){
            for(int j=0; j<grp; j++){
                if(j*cnt+i-1 < len)
                    ans[total++] = s[j*cnt+i-1];
                if(i != 1 && i != numRows){
                    if(j*cnt+i-1+2*(numRows-i) < len) ans[total++] = s[j*cnt+i-1+2*(numRows-i)];
                }
            }
        }
        ans[total] = 0;
        return string(ans);
    }
};
int main(){
    Solution s;
    cout<<s.convert("PAYPALISHIRING", 3)<<endl;
    cout<<s.convert("PAYPALISHIRING", 1)<<endl;
    cout<<s.convert("P", 1)<<endl;
    cout<<s.convert("PAYPALISHIRING", 4)<<endl;
    return 0;
}